src/main/java/org/openstreetmap/atlas/checks/validation/intersections/EdgeCrossingEdgeCheck.java

```diff
@@ -116,73 +90,25 @@ public boolean validCheckForObject(final AtlasObject object)
-        final Rectangle edgeBounds = edge.bounds();
-        final Optional<Long> edgeLayer = LayerTag.getTaggedValue(object);
-
-        // Retrieve crossing edges
-        final Atlas atlas = object.getAtlas();
-        final Iterable<Edge> crossingEdges = atlas.edgesIntersecting(edgeBounds,
-                // filter out the same edge, non-valid crossing edges and already flagged ones
-                crossingEdge -> edge.getIdentifier() != crossingEdge.getIdentifier()
-                        && isValidCrossingEdge(crossingEdge)
-                        && !this.isFlagged(generateAtlasObjectPairIdentifier(edge, crossingEdge)));
-
-        List<Edge> invalidEdges = null;
-
-        // Go through crossing items and collect invalid crossings
-        // NOTE: Due to way sectioning same OSM way could be marked multiple times here. However,
-        // MapRoulette will display way-sectioned edges in case there is an invalid crossing.
-        // Therefore, if an OSM way crosses another OSM way multiple times in separate edges, then
-        // each edge will be marked explicitly.
-        for (final Edge crossingEdge : crossingEdges)
-        {
-            final PolyLine crossingEdgeAsPolyLine = crossingEdge.asPolyLine();
-            final Optional<Long> crossingEdgeLayer = LayerTag.getTaggedValue(crossingEdge);
-            final Set<Location> intersections = edgeAsPolyLine
-                    .intersections(crossingEdgeAsPolyLine);
+        // Use BFS to gather all connected invalid crossings
+        final Set<Edge> invalidEdges = this.getInvalidCrossingEdges((Edge) object);
 
-            // Check whether crossing edge can actually cross
-            for (final Location intersection : intersections)
-            {
-                // Add this set to flagged pairs to skip it next time
-                this.markAsFlagged(generateAtlasObjectPairIdentifier(edge, crossingEdge));
-
-                // Check if crossing is valid or not
-                if (canCross(edgeAsPolyLine, edgeLayer, crossingEdgeAsPolyLine, crossingEdgeLayer,
-                        intersection))
-                {
-                    continue;
-                }
-
-                if (invalidEdges == null)
-                {
-                    // Normally we expect 1 or 2 edges in the list
-                    invalidEdges = new LinkedList<>();
-                }
-
-                invalidEdges.add(crossingEdge);
-            }
-        }
-
-        if (invalidEdges != null)
+        if (invalidEdges.size() > 1)
         {
             final CheckFlag newFlag = new CheckFlag(getTaskIdentifier(object));
+            this.markAsFlagged(object.getIdentifier());
             newFlag.addObject(object);
             newFlag.addInstruction(this.getLocalizedInstruction(0, object.getOsmIdentifier()));
 
             invalidEdges.forEach(invalidEdge ->
             {
+                this.markAsFlagged(invalidEdge.getIdentifier());
```

Comment:
where do you check to see if the edges you're working on are flagged?

Reply:
In the `isValidCrossingEdge` method on line 132.